controllers/machinepool_controller_noderef.go

```diff
@@ -0,0 +1,202 @@
+		return nil
+	}
+
+	// Check that Cluster isn't nil.
+	if cluster == nil {
+		logger.V(2).Info("MachinePool doesn't have a linked cluster, won't assign NodeRef")
+		return nil
+	}
+
+	logger = logger.WithValues("cluster", cluster.Name)
+
+	// Check that the MachinePool has valid ProviderIDs.
+	if len(machinepool.Spec.ProviderIDs) == 0 {
+		logger.Info("MachinePool doesn't have valid ProviderIDs yet")
+		return nil
+	}
+
+	clusterClient, err := remote.NewClusterClient(ctx, r.Client, cluster, r.scheme)
+	if err != nil {
+		return err
+	}
+
+	err = r.deleteRetiredNodes(ctx, clusterClient, machinepool.Status.NodeRefs, machinepool.Spec.ProviderIDs)
+	if err != nil {
+		return err
+	}
+
+	// Get the Node reference.
+	nodeRefsResult, err := r.getNodeReferences(ctx, clusterClient, machinepool.Spec.ProviderIDs)
+	if err != nil {
+		if err == ErrNodeNotFound {
+			return errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: 10 * time.Second},
+				"cannot assign NodeRefs to MachinePool %q in namespace %q, no matching Nodes", machinepool.Name, machinepool.Namespace)
+		}
+		logger.Error(err, "Failed to assign NodeRef")
+		r.recorder.Event(machinepool, apicorev1.EventTypeWarning, "FailedSetNodeRef", err.Error())
+		return err
+	}
+
+	machinepool.Status.AvailableReplicas = int32(nodeRefsResult.available)
+	machinepool.Status.ReadyReplicas = int32(nodeRefsResult.ready)
+	machinepool.Status.NodeRefs = nodeRefsResult.references
+	logger.Info("Set MachinePools's NodeRefs", "noderefs", machinepool.Status.NodeRefs)
+	r.recorder.Event(machinepool, apicorev1.EventTypeNormal, "SuccessfulSetNodeRef", fmt.Sprintf("%+v", machinepool.Status.NodeRefs))
+
+	if machinepool.Status.ReadyReplicas == 0 || len(nodeRefsResult.references) != int(machinepool.Status.ReadyReplicas) {
+		return errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: 10 * time.Second},
+			"NodeRefs != ReadyReplicas [%q != %q] for MachinePool %q in namespace %q", len(nodeRefsResult.references), machinepool.Status.ReadyReplicas, machinepool.Name, machinepool.Namespace)
+	}
+	return nil
+}
+
+// deleteRetiredNodes deletes nodes that don't have a corresponding ProviderID in Spec.ProviderIDs.
+// A MachinePool infrastucture provider indicates an instance in the set has been deleted by
+// removing its ProviderID from the slice.
+func (r *MachinePoolReconciler) deleteRetiredNodes(ctx context.Context, client client.Client, nodeRefs []apicorev1.ObjectReference, providerIDs []string) error {
+	logger := r.Log.WithValues("providerIDs", len(providerIDs))
+	nodeRefsMap := make(map[string]*apicorev1.Node, len(nodeRefs))
+	for _, nodeRef := range nodeRefs {
+		node := &corev1.Node{}
+		err := client.Get(ctx, types.NamespacedName{Name: nodeRef.Name}, node)
+		if err != nil {
+			return err
+		}
+
+		nodeProviderID, err := noderefutil.NewProviderID(node.Spec.ProviderID)
+		if err != nil {
+			logger.Error(err, "Failed to parse ProviderID", "node", node.Name)
```

Comment:
This seems more of a warning/info? We continue the iteration, and I don't think users can take any actions